flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/IndexFormatter.java

```diff
@@ -0,0 +1,235 @@
+import java.util.regex.Pattern;
+
+/**
+ * User can use {@link IndexFormatter} to format index value from index pattern.
+ *
+ * <p>If the index is a dynamic index, the index pattern include general pattern like
+ * 'connector.index'='my-index-{item}' and time extract pattern like 'connector.index'='my-index-{log_ts|yyyy-MM-dd}'.
+ *
+ * <p>For general pattern:
+ * 'item' the index field comes from any type column.
+ *
+ * <p>For time extract pattern:
+ * 'log_ts' is the index field comes from a varchar/timestamp column.
+ * 'yyyy-MM-dd' is the date format follows the {@link SimpleDateFormat} syntax.
+ * '{log_ts|yyyy-MM-dd}' is the time extract pattern for index in a dynamic index pattern.
+ */
+@Internal
+public class IndexFormatter implements Serializable {
+
+	private static final Pattern dynamicIndexPattern = Pattern.compile(".*\\{.+\\}.*");
+	private static final Pattern dynamicIndexTimeExtractPattern = Pattern.compile(".*\\{.+\\|.*\\}.*");
+	private static final String DEFAULT_FORMAT = "yyyy-MM-dd";
+
+	private String index;
+	private boolean dynamicIndexEnabled;
+	private boolean dynamicIndexTimeExtractEnabled;
+	private String dynamicIndexPatternStr;
+
+	private int indexFieldPos;
+	private TypeInformation indexFieldType;
+	private SimpleDateFormat dateFormat;
+
+	IndexFormatter(String index, String[] fieldNames, TypeInformation<?>[] fieldTypes) {
+		this.index = index;
+		this.dynamicIndexEnabled = checkDynamicIndexEnabled();
+		this.dynamicIndexPatternStr = extractDynamicIndexPatternStr();
+		this.dynamicIndexTimeExtractEnabled = checkDynamicIndexTimeExtractEnabled();
+		this.indexFieldPos = extractIndexFieldPos(fieldNames);
+		this.indexFieldType = extractIndexFieldType(fieldTypes);
+		this.dateFormat = extractDateFormat();
+	}
+
+	/**
+	 * Builder for {@link IndexFormatter}.
+	 */
+	public static class Builder {
+		private String index;
+		private String[] fieldNames;
+		private TypeInformation<?>[] fieldTypes;
+
+		public Builder index(String index) {
+			this.index = index;
+			return this;
+		}
+
+		public Builder fieldNames(String[] fieldNames) {
+			this.fieldNames = fieldNames;
+			return this;
+		}
+
+		public Builder fieldTypes(TypeInformation<?>[] fieldTypes) {
+			this.fieldTypes = fieldTypes;
+			return this;
+		}
+
+		public IndexFormatter build() {
+			return new IndexFormatter(index, fieldNames, fieldTypes);
+		}
+	}
+
+	/**
+	 * Creates a new builder for {@link IndexFormatter}.
+	 */
+	public static Builder builder() {
+		return new Builder();
+	}
+
+	public boolean isDynamicIndexEnabled() {
+		return dynamicIndexEnabled;
+	}
+
+	/**
+	 * Return dynamic index if dynamic index is enabled, else return the static index.
+	 */
+	public String getFormattedIndex(Row row) {
+		if (!isDynamicIndexEnabled()) {
+			return index;
+		}
+		Object indexFiled = row.getField(indexFieldPos);
+		String indexFiledValueStr = indexFiled.toString();
+
+		if (dynamicIndexTimeExtractEnabled) {
+			if (indexFieldType == Types.LONG) {
```

Comment:
We already know `indexFieldType`, could we do this switch before runtime, like `converters` in json-format? Then we could avoid this `if-else` for each record.

Reply:
nice tips